File: sdk/python/provenance_client/interfaces/gateway.py
from typing import Any, Callable, Protocol

from ..core.models import Decision, ExecutionResult


class ProvenanceGatewayProtocol(Protocol):
    def guard(
        self,
        action: str | None = None,
        *,
        session_id: str,
        **kwargs,
    ) -> Callable[[Any], Any]: ...

    def execute(
        self,
        action: str,
        parameters: dict,
        *,
        session_id: str | None = None,
        decision: Decision | None = None,
    ) -> ExecutionResult: ...

    async def async_execute(
        self,
        action: str,
        parameters: dict,
        *,
        session_id: str | None = None,
        decision: Decision | None = None,
    ) -> ExecutionResult: ...

File: sdk/typescript/provenance_client/interfaces/gateway.ts
import { Decision, ExecutionResult } from "../core/models";

export interface ProvenanceGatewayProtocol {
  guard(
    action?: string,
    options?: {
      sessionId?: string;
      decision?: Decision;
      raiseOnBlock?: boolean;
    },
  ): <T extends (...args: any[]) => any>(func: T) => T;

  execute(
    action: string,
    parameters: Record<string, any>,
    options?: {
      sessionId?: string;
      decision?: Decision;
    },
  ): ExecutionResult;

  asyncExecute(
    action: string,
    parameters: Record<string, any>,
    options?: {
      sessionId?: string;
      decision?: Decision;
    },
  ): Promise<ExecutionResult>;
}
